### ISSUE TITLE
loss.backward() very slow on tpu

### ISSUE THREAD

**Issue by christinaknudsen** [pytorch/xla#2655]
## ❓ Questions and Help
Hi!

I am finetunig a Transformers model, and tried to run the code on tpu to speed things up. However, the loss.backward() function seems to take forever (>10 min). I'm afraid some of the computations are done on the cpu. What might be the problem?

I run the code in google colab. Here are some details:

`
    
    VERSION = "nightly"  #@param ["nightly", "20200516"]  # or YYYYMMDD format
    
    !curl https://raw.githubusercontent.com/pytorch/xla/master/contrib/scripts/env-setup.py -o pytorch-xla-env-setup.py
    !python pytorch-xla-env-setup.py --version $VERSION

    os.environ['XLA_USE_BF16']="1"
    os.environ['XLA_TENSOR_ALLOCATOR_MAXSIZE'] = '100000000'
    !export XLA_USE_BF16=1

    mx = AutoModelWithLMHead.from_pretrained(
        args.model_name_or_path,
        from_tf=False,
        config=config,
        cache_dir=args.cache_dir,
    )

    def main():
        device = xm.xla_device()
        model=mx.to(device)

        train_sampler = DistributedSampler(
                  train_dataset,
                  num_replicas=xm.xrt_world_size(),
                  rank=xm.get_ordinal(),
                  shuffle=True)

        train_dataloader = DataLoader(
                    train_dataset,
                    batch_size= args.train_per_tpu_batch_size *xm.xrt_world_size(),
                    sampler=train_sampler,
                    drop_last=True,
                    num_workers=4,
                    collate_fn = collate
                    )
        no_decay = ["bias", "LayerNorm.weight"]
        optimizer_grouped_parameters = [
            {"params": [p for n, p in model.named_parameters() if not any(nd in n for nd in no_decay)],
            "weight_decay": args.weight_decay,
            },
            {"params": [p for n, p in model.named_parameters() if any(nd in n for nd in no_decay)],
            "weight_decay": 0.0},
        ]
        optimizer = AdamW(optimizer_grouped_parameters, lr=args.learning_rate, eps=args.adam_epsilon)
        scheduler = get_linear_schedule_with_warmup(
            optimizer, num_warmup_steps=args.warmup_steps, num_training_steps=num_train_steps
        )
        for _ in range(args.num_train_epochs):
            para_loader = pl.ParallelLoader(train_dataloader, [device])
            global_step, tr_loss = trainer(para_loader.per_device_loader(device), model, optimizer, device, scheduler)

    def trainer(dataloader, model,optimizer,device,scheduler)
        for step, batch in enumerate(data_loader):
            model.zero_grad()
            inputs, labels = (batch, batch)
            inputs = inputs.to(device)
            labels = labels.to(device)
            model.train()
            outputs = model(inputs, labels=labels)
            loss = outputs[0]
            loss.backward()
            tr_loss += loss.item()
            xm.optimizer_step(optimizer)
            scheduler.step()

    def _mp_fn(rank, flags):
        a = main()

    FLAGS={}
    start_time = time.time()
    xmp.spawn(_mp_fn, args=(FLAGS,), nprocs=8, start_method='fork')
`
This is what I got from the metric report:

Metric: CompileTime
  TotalSamples: 26
  Accumulator: 07m60s991ms747.685us
  ValueRate: 772ms008.427us / second
  Rate: 0.0477921 / second
  Percentiles: 1%=192ms721.273us; 5%=03s212ms401.694us; 10%=08s716ms887.597us; 20%=10s092ms102.342us; 50%=16s340ms436.548us; 80%=23s048ms986.235us; 90%=25s500ms218.199us; 95%=25s626ms475.380us; 99%=26s881ms519.201us
Metric: DeviceLockWait
  TotalSamples: 28
  Accumulator: 097.952us
  ValueRate: 000.179us / second
  Rate: 0.0510697 / second
  Percentiles: 1%=001.935us; 5%=001.992us; 10%=002.008us; 20%=002.110us; 50%=003.414us; 80%=004.346us; 90%=005.689us; 95%=005.711us; 99%=007.577us
Metric: ExecuteTime
  TotalSamples: 26
  Accumulator: 04s997ms628.876us
  ValueRate: 007ms342.894us / second
  Rate: 0.0477691 / second
  Percentiles: 1%=013ms217.352us; 5%=014ms419.637us; 10%=031ms457.717us; 20%=058ms364.298us; 50%=149ms928.505us; 80%=260ms040.681us; 90%=277ms947.144us; 95%=278ms972.906us; 99%=335ms850.157us
Metric: InboundData
  TotalSamples: 26
  Accumulator: 3.80GB
  ValueRate: 7.30MB / second
  Rate: 0.0488075 / second
  Percentiles: 1%=2.00B; 5%=2.00B; 10%=162.00MB; 20%=162.00MB; 50%=162.00MB; 80%=162.00MB; 90%=162.00MB; 95%=162.00MB; 99%=162.00MB
Metric: InputOutputAliasCount
  TotalSamples: 26
  Accumulator: 210.00
  ValueRate: 0.40 / second
  Rate: 0.0500392 / second
  Percentiles: 1%=1.00; 5%=1.00; 10%=1.00; 20%=1.00; 50%=2.00; 80%=2.00; 90%=2.00; 95%=2.00; 99%=173.00
Metric: IrValueTensorToXlaData
  TotalSamples: 185
  Accumulator: 59s948ms810.782us
  ValueRate: 105ms028.699us / second
  Rate: 0.329619 / second
  Percentiles: 1%=971.366us; 5%=005ms330.099us; 10%=010ms280.273us; 20%=024ms254.924us; 50%=057ms332.349us; 80%=149ms132.659us; 90%=01s128ms605.598us; 95%=02s069ms349.285us; 99%=03s010ms297.826us
Metric: OutboundData
  TotalSamples: 206
  Accumulator: 3.16GB
  ValueRate: 5.74MB / second
  Rate: 0.365302 / second
  Percentiles: 1%=2.00B; 5%=2.00B; 10%=1.50KB; 20%=1.50KB; 50%=4.50KB; 80%=4.50MB; 90%=81.00MB; 95%=162.00MB; 99%=162.00MB
Metric: ReleaseDataHandlesTime
  TotalSamples: 38
  Accumulator: 257ms000.381us
  ValueRate: 469.035us / second
  Rate: 0.0693514 / second
  Percentiles: 1%=001ms042.074us; 5%=001ms079.476us; 10%=001ms191.167us; 20%=001ms238.119us; 50%=002ms560.938us; 80%=014ms812.687us; 90%=026ms308.251us; 95%=028ms945.489us; 99%=028ms269.511us
Metric: TensorsGraphSize
  TotalSamples: 26
  Accumulator: 43095.00
  ValueRate: 79.22 / second
  Rate: 0.047792 / second
  Percentiles: 1%=147.00; 5%=292.00; 10%=437.00; 20%=853.00; 50%=1802.00; 80%=2478.00; 90%=2793.00; 95%=2898.00; 99%=3003.00
Metric: TransferFromServerTime
  TotalSamples: 26
  Accumulator: 01m01s310ms131.419us
  ValueRate: 115ms092.119us / second
  Rate: 0.0488075 / second
  Percentiles: 1%=006ms573.035us; 5%=009ms753.050us; 10%=02s683ms770.772us; 20%=02s826ms130.774us; 50%=02s151ms068.869us; 80%=03s094ms740.098us; 90%=04s905ms632.493us; 95%=04s104ms528.303us; 99%=04s222ms886.906us
Metric: TransferToServerTime
  TotalSamples: 206
  Accumulator: 59s429ms656.071us
  ValueRate: 106ms864.883us / second
  Rate: 0.366964 / second
  Percentiles: 1%=002ms738.097us; 5%=002ms400.462us; 10%=008ms035.623us; 20%=018ms255.764us; 50%=053ms168.961us; 80%=143ms896.739us; 90%=01s021ms159.368us; 95%=02s058ms501.411us; 99%=03s856ms341.418us
Metric: TransferToServerTransformTime
  TotalSamples: 206
  Accumulator: 03s346ms256.572us
  ValueRate: 006ms933.948us / second
  Rate: 0.365302 / second
  Percentiles: 1%=063.485us; 5%=077.473us; 10%=089.022us; 20%=113.860us; 50%=348.909us; 80%=010ms562.090us; 90%=051ms242.880us; 95%=091ms607.269us; 99%=163ms426.983us
Counter: CreateCompileHandles
  Value: 26
Counter: CreateDataHandles
  Value: 422
Counter: CreateXlaTensor
  Value: 3631
Counter: DestroyDataHandles
  Value: 209
Counter: DestroyXlaTensor
  Value: 2965
Counter: DeviceDataCacheMiss
  Value: 18
Counter: MarkStep
  Value: 1
Counter: ReleaseDataHandles
  Value: 209
Counter: UncachedCompile
  Value: 26
Counter: XRTAllocateFromTensor_Empty
  Value: 24
Counter: XrtCompile_Empty
  Value: 32
Counter: XrtExecuteChained_Empty
  Value: 32
Counter: XrtExecute_Empty
  Value: 32
Counter: XrtMemoryInfo_Empty
  Value: 32
Counter: XrtRead_Empty
  Value: 32
Counter: XrtReleaseAllocationHandle_Empty
  Value: 32
Counter: XrtReleaseCompileHandle_Empty
  Value: 32
Counter: XrtSessionCount
  Value: 4
Counter: XrtSubTuple_Empty
  Value: 32
Counter: aten::_local_scalar_dense
  Value: 1
Counter: aten::conj_out
  Value: 12
Counter: aten::view_as_real
  Value: 12
Counter: xla::_copy_from
  Value: 175
Counter: xla::_log_softmax
  Value: 1
Counter: xla::_log_softmax_backward_data
  Value: 1
Counter: xla::_s_where
  Value: 24
Counter: xla::_softmax
  Value: 12
Counter: xla::_softmax_backward_data
  Value: 12
Counter: xla::_unsafe_view
  Value: 25
Counter: xla::add
  Value: 98
Counter: xla::add_
  Value: 298
Counter: xla::addcdiv_
  Value: 149
Counter: xla::addcmul
  Value: 25
Counter: xla::addcmul_
  Value: 149
Counter: xla::addmm
  Value: 48
Counter: xla::arange_out
  Value: 1
Counter: xla::as_strided
  Value: 173
Counter: xla::bernoulli_
  Value: 37
Counter: xla::bmm
  Value: 72
Counter: xla::cat
  Value: 12
Counter: xla::div
  Value: 24
Counter: xla::div_
  Value: 37
Counter: xla::embedding
  Value: 2
Counter: xla::embedding_dense_backward
  Value: 2
Counter: xla::empty
  Value: 536
Counter: xla::empty_strided
  Value: 173
Counter: xla::expand
  Value: 84
Counter: xla::fill_
  Value: 1
Counter: xla::index_select
  Value: 2
Counter: xla::mm
  Value: 99
Counter: xla::mul
  Value: 306
Counter: xla::mul_
  Value: 298
Counter: xla::native_batch_norm
  Value: 25
Counter: xla::native_batch_norm_backward
  Value: 25
Counter: xla::nll_loss_backward
  Value: 1
Counter: xla::nll_loss_forward
  Value: 1
Counter: xla::permute
  Value: 96
Counter: xla::pow
  Value: 24
Counter: xla::select
  Value: 12
Counter: xla::slice
  Value: 53
Counter: xla::split
  Value: 12
Counter: xla::sqrt
  Value: 149
Counter: xla::stack
  Value: 12
Counter: xla::sum
  Value: 99
Counter: xla::t
  Value: 100
Counter: xla::tanh
  Value: 12
Counter: xla::tanh_backward
  Value: 12
Counter: xla::transpose
  Value: 60
Counter: xla::unsqueeze
  Value: 1
Counter: xla::view
  Value: 645
Counter: xla::zero_
  Value: 312
Metric: XrtAllocateFromTensor
  TotalSamples: 1596
  Accumulator: 37s646ms420.741us
  Mean: 035ms819.128us
  StdDev: 104ms580.870us
  Rate: 1.84943 / second
  Percentiles: 25%=514.972us; 50%=808.954us; 80%=003ms259.539us; 90%=065ms084.694us; 95%=335ms549.324us; 99%=493ms515.590us
Metric: XrtCompile
  TotalSamples: 172
  Accumulator: 54m29s610ms567.523us
  Mean: 19s004ms543.997us
  StdDev: 09s058ms036.388us
  Rate: 0.314054 / second
  Percentiles: 25%=13s454ms544.155us; 50%=19s735ms202.926us; 80%=25s785ms116.934us; 90%=33s898ms462.353us; 95%=36s076ms244.486us; 99%=39s285ms453.683us
Metric: XrtExecute
  TotalSamples: 172
  Accumulator: 33s800ms995.578us
  Mean: 191ms697.649us
  StdDev: 116ms060.946us
  Rate: 0.314053 / second
  Percentiles: 25%=096ms970.272us; 50%=186ms324.192us; 80%=295ms643.447us; 90%=352ms645.310us; 95%=401ms180.462us; 99%=440ms566.327us
Metric: XrtExecutorEvict
  TotalSamples: 0
  Accumulator: nanB
  Mean: nanB
  StdDev: nanB
  Percentiles: 
Metric: XrtReadLiteral
  TotalSamples: 246
  Accumulator: 05m39s743ms153.167us
  Mean: 01s133ms102.249us
  StdDev: 902ms860.228us
  Rate: 0.457093 / second
  Percentiles: 25%=280.789us; 50%=02s507ms762.211us; 80%=02s913ms176.790us; 90%=02s088ms548.834us; 95%=02s364ms055.592us; 99%=03s671ms844.821us
Metric: XrtReleaseAllocation
  TotalSamples: 258
  Accumulator: 014ms110.464us
  Mean: 054.692us
  StdDev: 040.823us
  Rate: 0.46827 / second
  Percentiles: 25%=028.839us; 50%=046.897us; 80%=064.627us; 90%=090.488us; 95%=137.983us; 99%=262.009us

**Comment by JackCaoG** [comment]
I think the main problem is there are 2 aten ops don't have xla lowering
```
Counter: aten::conj_out
Value: 12
Counter: aten::view_as_real
Value: 12
```

This is why we have 26 compiles which takes 26 mins, where execute only take 4 seconds
```
Metric: CompileTime
TotalSamples: 26
Accumulator: 07m60s991ms747.685us
ValueRate: 772ms008.427us / second
Rate: 0.0477921 / second
Percentiles: 1%=192ms721.273us; 5%=03s212ms401.694us; 10%=08s716ms887.597us; 20%=10s092ms102.342us; 50%=16s340ms436.548us; 80%=23s048ms986.235us; 90%=25s500ms218.199us; 95%=25s626ms475.380us; 99%=26s881ms519.201us


Metric: ExecuteTime
TotalSamples: 26
Accumulator: 04s997ms628.876us
ValueRate: 007ms342.894us / second
Rate: 0.0477691 / second
Percentiles: 1%=013ms217.352us; 5%=014ms419.637us; 10%=031ms457.717us; 20%=058ms364.298us; 50%=149ms928.505us; 80%=260ms040.681us; 90%=277ms947.144us; 95%=278ms972.906us; 99%=335ms850.157us
```

Do you know where are these 2 ops coming from and if they are really needed?

**Comment by rabeehkarimimahabadi** [comment]
indeed this is very slow for me too.

**Comment by Wikidepia** [comment]
Same, its very slow here too. I use nightly version

**Comment by christinaknudsen** [comment]
Hi!

Thanks for your response. I don't know where they might come from... Do you have any suggestions?

Christina

**Comment by taylanbil** [comment]
if you have `aten::________`'s in your workload, you can catch those by running on 1 core, inserting a breakpoint at various places of your model's forward, and checking the metrics report via something like:

```python
def hasaten():
  report = met.metrics_report().split('\n')
  for line in report:
    return "Counter: aten" in line
  return False
```

**Comment by taylanbil** [comment]
This issue is orobably resolved per https://github.com/pytorch/xla/issues/2688, could you check with latest version of nightly?

**Comment by christinaknudsen** [comment]
Hi!

Yes, you are right, no more aten's, and the backward is really fast. Thank you!
